src/test/kotlin/io/csbroker/apiserver/controller/IntegrationTest.kt

```diff
@@ -0,0 +1,139 @@
+    private lateinit var authTokenProvider: AuthTokenProvider
+
+    fun request(
+        method: HttpMethod,
+        url: String,
+        body: Any? = null,
+        headers: MultiValueMap<String, String> = LinkedMultiValueMap(),
+        params: MultiValueMap<String, String> = LinkedMultiValueMap(),
+        isAdmin: Boolean = false,
+    ): ResultActions {
+        val req = MockMvcRequestBuilders.request(method, url)
+            .headers(HttpHeaders(headers.also { it.add("Authorization", "Bearer ${createToken(isAdmin)}") }))
+            .params(params)
+            .contentType(MediaType.APPLICATION_JSON)
+            .accept(MediaType.APPLICATION_JSON)
+
+        body?.let {
+            req.content(objectMapper().writeValueAsString(it))
+        }
+
+        return mockMvc.perform(req)
+    }
+
+    fun <T> save(entity: T): T = runWithEntityManager(readonly = false) {
+        it.persist(entity)
+        entity
+    }
+
+    fun <T> findOne(queryString: String, params: Map<String, Any> = emptyMap()): T = runWithEntityManager {
+        val query = it.createQuery(queryString, Any::class.java)
+        setParams(query, params)
+        query.singleResult as T
+    }
+
+    fun <T> findAll(queryString: String, params: Map<String, Any> = emptyMap()): List<T> = runWithEntityManager {
+        val query = it.createQuery(queryString, Any::class.java)
+        setParams(query, params)
+        query.resultList as List<T>
+    }
+
+    fun <T> findById(id: Long, clazz: Class<T>): T = runWithEntityManager {
+        it.find(clazz, id)
+    }
+
+    private fun setParams(query: Query, params: Map<String, Any>) {
+        params.forEach { (key, value) ->
+            query.setParameter(key, value)
+        }
+    }
+
+    private fun <T> runWithEntityManager(readonly: Boolean = true, block: (em: EntityManager) -> T): T {
+        val em = entityManagerFactory.createEntityManager()
+        val result = if (readonly) {
+            runCatching {
+                block(em)
+            }
+        } else {
+            val transaction = em.transaction
+            transaction.begin()
+            runCatching {
+                block(em)
+            }.onFailure {
+                transaction.rollback()
+            }.onSuccess {
+                transaction.commit()
+            }
+        }
+        em.close()
+        return result.getOrThrow()
+    }
+
+    private fun createToken(isAdmin: Boolean): String {
+        val users = findAll<User>("SELECT u FROM User u where u.email = 'test@csbroker.io'")
+        val user = if (users.isNotEmpty()) {
+            users.first()
+        } else {
+            save(
+                User(
+                    email = "test@csbroker.io",
+                    username = "test",
+                    password = "test",
+                    role = if (isAdmin) Role.ROLE_ADMIN else Role.ROLE_USER,
+                    providerType = ProviderType.LOCAL,
+                ),
+            )
+        }
```

Comment:
만약 isAdmin이 true 인데  user를 실수로 admin이 아니게 작성해서 first로 나온 유저의 권한이 의도하지 않게 나올 가능성이 있어보입니다 🤔
```suggestion
        val role = if (isAdmin) Role.ROLE_ADMIN else Role.ROLE_USER
        val user = users.find { it.role == role } ?: save(
            User(
                email = "test@csbroker.io",
                username = "test",
                password = "test",
                role = role,
                providerType = ProviderType.LOCAL,
            ),
        )
```